pkgs/native_toolchain_c/lib/src/cbuilder/linker_options.dart

```diff
@@ -85,38 +88,43 @@ class LinkerOptions {
 }
 
 extension LinkerOptionsExt on LinkerOptions {
-  /// The flags for the specified [linker], which are inserted _before_ the
-  /// sources.
-  ///
-  /// This is mainly used for the whole-archive ... no-whole-archive
-  /// trick, which includes all symbols when linking object files.
-  ///
-  /// Throws if the [linker] is not supported.
-  Iterable<String> preSourcesFlags(Tool linker, Iterable<String> sourceFiles) =>
-      _toLinkerSyntax(
-        linker,
-        sourceFiles.any((source) => source.endsWith('.a')) ||
-                _wholeArchiveSandwich
-            ? ['--whole-archive']
-            : [],
-      );
-
-  /// The flags for the specified [linker], which are inserted _after_ the
-  /// sources.
-  ///
-  /// This is mainly used for the whole-archive ... no-whole-archive
-  /// trick, which includes all symbols when linking object files.
-  ///
-  /// Throws if the [linker] is not supported.
-  Iterable<String> postSourcesFlags(
-    Tool linker,
+  /// Takes [sourceFiles] and turns it into flags for the compiler driver while
+  /// considering the current [LinkerOptions].
+  Iterable<String> sourceFilesToFlags(
+    Tool tool,
     Iterable<String> sourceFiles,
-  ) => _toLinkerSyntax(linker, [
-    ..._linkerFlags,
-    if (gcSections) '--gc-sections',
-    if (linkerScript != null) '--version-script=${linkerScript!.toFilePath()}',
-    if (sourceFiles.any((source) => source.endsWith('.a')) ||
-        _wholeArchiveSandwich)
-      '--no-whole-archive',
-  ]);
+    OS targetOS,
+  ) {
+    final includeAllSymbols = _symbolsToKeep == null;
+
+    if (targetOS == OS.macOS) {
```

Comment:
Maybe this should be written as a switch on target OS (which then has both the Android and Linux cases got to the same block). Early return for MacOS and combining all other OSes seems somewhat weird.